dcicutils/project_utils.py

```diff
@@ -0,0 +1,453 @@
+
+    PYPROJECT_NAME | NAME        |REPO_NAME    |PYPI_NAME    |PACKAGE_NAME  |PRETTY_NAME  |APP_NAME  |APP_PRETTY_NAME
+    ---------------+-------------+-------------+-------------+--------------+-------------+----------+----------------
+    dcicsnovault   |snovault     |snovault     |snovault     |snovault      |Snovault     |snovault  |Snovault
+    encoded-core   |encoded-core |encoded-core |encoded-core |encoded-core  |Encoded Core |core      |Core
+    encoded        |cgap-portal  |cgap-portal  |None         |encoded       |CGAP Portal  |cgap      |CGAP
+    encoded        |fourfront    |fourfront    |None         |encoded       |Fourfront    |fourfront |Fourfront
+    encoded        |smaht-portal |smaht-portal |None         |encoded       |SMaHT Portal |smaht     |SMaHT
+
+     The registered name is the one used with the ProjectRegistry.register() decorator.
+    """
+
+    # This contains only the initial spec and is NOT used after initialization. See .identity (and ._identity).
+    # It is declared in this way so that actual declarations of project classes don't show up as syntax errors.
+    IDENTITY: Optional[dict] = None
+
+    IDENTITY_CLASS: Type[ProjectIdentity] = ProjectIdentity
+
+    _identity: Optional[ProjectIdentity] = None  # Class-level default, overridden in instances as a ProjectIdentity
+
+    _PROJECT_REGISTRY_CLASS = None
+
+    @classproperty
+    def PROJECT_REGISTRY_CLASS(cls):  # noQA - PyCharm thinks this should use 'self'
+        registry_class: Optional[Type[ProjectRegistry]] = cls._PROJECT_REGISTRY_CLASS
+        if registry_class is None:
+            raise ValueError(f"Cannot compute {cls.__name__}.PROJECT_REGISTRY_CLASS"
+                             f" because {cls.__name__}.declare_project_registry_class(...) has not been done.")
+        return registry_class
+
+    @classmethod
+    def declare_project_registry_class(cls, registry_class: Type):
+        """
+        This function is needed to complete a circularity where project classes need to know the appropriate
+        registry class and registry classes need to know the appropriate project class, so by convention when
+        the registry class (which is defined after the project class) is fully defined, this method on the project
+        class should be called with the registry class.
+        """
+        if not isinstance(registry_class, type) or not issubclass(registry_class, ProjectRegistry):
+            raise ValueError(f"The registry_class, {registry_class!r}, is not a subclass of ProjectRegistry.")
+        registry_class: Type[ProjectRegistry]
+        cls._PROJECT_REGISTRY_CLASS = registry_class
+
+    def __init__(self):
+        self._identity: ProjectIdentity = self.IDENTITY_CLASS(**self.IDENTITY)
+
+    def __str__(self):
+        return f"<{self.__class__.__name__} {id(self):x}>"
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}()"
```

Comment:
Are these representations really helpful? You might consider including more information?

Reply:
The `__repr__` one I just did because I was in there. The `__str__` one just keeps some really long ugly names from appearing in testing. All one really needs to know is the class. Even the `id` isn't really always needed, but it's useful in testing when I'm making more than one instance so I don't confuse myself.

Reply:
I'll have __repr__ call __str__